FWCore/Framework/interface/maker/Worker.h

```diff
@@ -690,7 +699,12 @@ namespace edm {
                        ModuleCallingContext const* mcc,
                        Arg::Context const* context) {
         ModuleSignalSentry<Arg> cpp(actReg, context, mcc);
-        return iWorker->implDoBegin(info, mcc);
+        if (actReg) {
+          Arg::preModuleSignal(actReg, context, mcc);
+        }
+        auto returnValue = iWorker->implDoBegin(info, mcc);
+        cpp.postModuleSignal();
+        return returnValue;
```

Comment:
1. I moved the pre module signal out of the ```ModuleSignalSentry``` constructor, because if the pre module signal throws inside the constructor then the destructor is not called and the post module signal is not emitted. (Point 3: we want "If the pre module signal throws, then module does not execute but the post signal executes anyway").
2. If an exception was thrown, then the destructor calls the post module signal. But if an exception was not thrown, then the function ```postModuleSignal``` calls the post module signal. The post module signal is only emitted once, not by both.
3. Note that the function ```postModuleSignal``` sets a_ to null if it is called and that causes the destructor to do nothing when it is called.
4. The destructor only calls the post module signal after an exception was already thrown in the module or the pre module signal. We ignore the exception in that case, because the earlier exception is the one we want reported.
5. We don't want exceptions ignored from the post module signal in ```postModuleSignal``` because it is the first exception in that case.

Reply:
Thanks, I think I got the logic now. So the explicit call to `postModuleSignal()` is necessary in order to have the "let exception propagate" vs. "ignore exception" behavior difference wrt. whether an exception was already thrown in this scope.

The repeated pattern to call the `preModuleSignal()` still bugs me a little bit, but after thinking some time of alternatives (such as nested try-catch in the `ModuleSignalSentry<T>` constructor, another Sentry class) I'm coming to the conclusion the repeated pattern would not be worth of addressing.

Reply:
I could define a function called ```preModuleSignal``` in the sentry. Then I could reduce the 3 lines that are repeated down to 1 line that is repeated. Is that better?

Reply:
I think that is better. I will do that unless you don't like it. Add the following function to ```ModuleSignalSentry```
```
      void preModuleSignal() {
        if (a_) {
          T::preModuleSignal(a_, context, moduleCallingContext_);
        }
      }
```

Reply:
I'm good with your suggestion of adding `preModuleSignal()` function to the sentry.